# **1. Import Library**

Pada tahap ini, Anda perlu mengimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning.

In [65]:
import pandas as pd
import google.colab as colab
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, f1_score
colab.drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# **2. Memuat Dataset dari Hasil Clustering**

Memuat dataset hasil clustering dari file CSV ke dalam variabel DataFrame.

In [66]:
cluster_df = pd.read_csv('/content/drive/MyDrive/Coding Camp DBS Foundation/Proyek/Dataset_inisiasi.csv')
cluster_df.head()

,price,processor_type,ssd_capacity,ram_system_memory,graphics_card,capacity,cluster
0,13716.90,Intel Core i5,512.0,16.0,AMD Radeon RX 550,512.0,4
1,13678.00,Intel Core i5,512.0,16.0,Nvidia Geforce GT 740,0.0,2
2,12857.74,Intel Core i5,512.0,16.0,AMD Radeon RX 580,512.0,0
3,20799.00,Intel Core i5,256.0,32.0,AMD Radeon RX 580,256.0,0
4,12834.62,Intel Core i5,512.0,16.0,AMD Radeon RX 550,512.0,4


## Encode Label

In [67]:
# Create a LabelEncoder instance
encoder = LabelEncoder()

# Fit and transform each column separately
cluster_df['graphics_card'] = encoder.fit_transform(cluster_df['graphics_card'])
cluster_df['processor_type'] = encoder.fit_transform(cluster_df['processor_type'])

cluster_df.head()

,price,processor_type,ssd_capacity,ram_system_memory,graphics_card,capacity,cluster
0,13716.90,7,512.0,16.0,5,512.0,4
1,13678.00,7,512.0,16.0,43,0.0,2
2,12857.74,7,512.0,16.0,9,512.0,0
3,20799.00,7,256.0,32.0,9,256.0,0
4,12834.62,7,512.0,16.0,5,512.0,4


# **3. Data Splitting**

Tahap Data Splitting bertujuan untuk memisahkan dataset menjadi dua bagian: data latih (training set) dan data uji (test set).

In [68]:
X = cluster_df.drop(['cluster'], axis=1)
y = cluster_df['cluster']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Jumlah data latih:", X_train.shape[0])
print("Jumlah data uji:", X_test.shape[0])


Jumlah data latih: 2053
Jumlah data uji: 514


# **4. Membangun Model Klasifikasi**


## **a. Membangun Model Klasifikasi**

Setelah memilih algoritma klasifikasi yang sesuai, langkah selanjutnya adalah melatih model menggunakan data latih.

Berikut adalah rekomendasi tahapannya.
1. Pilih algoritma klasifikasi yang sesuai, seperti Logistic Regression, Decision Tree, Random Forest, atau K-Nearest Neighbors (KNN).
2. Latih model menggunakan data latih.

In [69]:
model_random_forest = RandomForestClassifier(n_estimators=10,
                                             max_depth=5,
                                             min_samples_leaf=1,
                                             min_samples_split=2,
                                             max_features="sqrt",
                                             n_jobs=-1,
                                             random_state=42)
model_random_forest.fit(X_train, y_train)

model_logistic = LogisticRegression(penalty="l2",
                                    solver="lbfgs",
                                    C=0.1,
                                    max_iter=1000,
                                    tol=0.0001,
                                    random_state=42,
                                    n_jobs=-1)
model_logistic.fit(X_train, y_train)

LogisticRegression(C=0.1, max_iter=1000, n_jobs=-1, random_state=42)

Tulis narasi atau penjelasan algoritma yang Anda gunakan.

## **b. Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Lakukan prediksi menggunakan data uji.
2. Hitung metrik evaluasi seperti Accuracy dan F1-Score (Opsional: Precision dan Recall).
3. Buat confusion matrix untuk melihat detail prediksi benar dan salah.

In [70]:
model_random_forest.predict(X_test)
model_logistic.predict(X_test)

model_random_forest.score(X_test, y_test)
model_logistic.score(X_test, y_test)

print("Model Random Forest")
print(classification_report(y_test, model_random_forest.predict(X_test)))
print(confusion_matrix(y_test, model_random_forest.predict(X_test)))
print("Accuracy Score : ", accuracy_score(y_test, model_random_forest.predict(X_test)))
print("F1 Score : ", f1_score(y_test, model_random_forest.predict(X_test), average='weighted'))

print("Model Logistic")
print(classification_report(y_test, model_logistic.predict(X_test)))
print(confusion_matrix(y_test, model_logistic.predict(X_test)))
print("Accuracy Score : ", accuracy_score(y_test, model_logistic.predict(X_test)))
print("F1 Score : ", f1_score(y_test, model_logistic.predict(X_test), average='weighted'))

Model Random Forest
              precision    recall  f1-score   support

           0       0.86      0.96      0.91        80
           1       0.99      1.00      1.00       147
           2       1.00      0.83      0.91         6
           3       1.00      0.99      0.99        92
           4       1.00      0.24      0.38        21
           5       1.00      0.93      0.96        14
           6       0.83      1.00      0.91         5
           7       0.97      0.97      0.97       116
           8       0.73      0.91      0.81        33

    accuracy                           0.94       514
   macro avg       0.93      0.87      0.87       514
weighted avg       0.95      0.94      0.94       514

[[ 77   0   0   0   0   0   0   0   3]
 [  0 147   0   0   0   0   0   0   0]
 [  0   0   5   0   0   0   1   0   0]
 [  0   0   0  91   0   0   0   1   0]
 [  9   0   0   0   5   0   0   1   6]
 [  0   1   0   0   0  13   0   0   0]
 [  0   0   0   0   0   0   5   0   0]
 [

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Tulis hasil evaluasi algoritma yang digunakan, jika Anda menggunakan 2 algoritma, maka bandingkan hasilnya.

## **c. Tuning Model Klasifikasi (Optional)**

Gunakan GridSearchCV, RandomizedSearchCV, atau metode lainnya untuk mencari kombinasi hyperparameter terbaik

In [72]:
logistick_param = {
    'penalty': ['l1', 'l2'],
    'C': [0.001, 0.01, 0.1, 1],
    'solver': ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga'],
    'max_iter': [100,200,300,400,500],
    'tol': [1e-5, 1e-4, 1e-3, 1e-2, 1e-1]
}

grid_logistic = GridSearchCV(estimator=LogisticRegression(), param_grid=logistick_param, cv=5)

model_logistic = grid_logistic.fit(X_train, y_train)
model_logistic.best_params_

print("Model Logistic")
print(classification_report(y_test, model_logistic.predict(X_test)))
print(confusion_matrix(y_test, model_logistic.predict(X_test)))
print("Accuracy Score : ", accuracy_score(y_test, model_logistic.predict(X_test)))
print("F1 Score : ", f1_score(y_test, model_logistic.predict(X_test), average='weighted'))


Streaming output truncated to the last 5000 lines.
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.htm

Model Logistic
              precision    recall  f1-score   support

           0       0.89      0.95      0.92        80
           1       0.93      0.99      0.96       147
           2       0.75      0.50      0.60         6
           3       0.99      0.97      0.98        92
           4       1.00      1.00      1.00        21
           5       1.00      0.43      0.60        14
           6       0.57      0.80      0.67         5
           7       0.96      0.92      0.94       116
           8       0.88      0.91      0.90        33

    accuracy                           0.94       514
   macro avg       0.89      0.83      0.84       514
weighted avg       0.94      0.94      0.93       514

[[ 76   0   0   0   0   0   0   3   1]
 [  0 145   1   0   0   0   1   0   0]
 [  0   3   3   0   0   0   0   0   0]
 [  0   1   0  89   0   0   0   1   1]
 [  0   0   0   0  21   0   0   0   0]
 [  0   6   0   0   0   6   2   0   0]
 [  0   1   0   0   0   0   4   0   0]
 [  6  

/usr/local/lib/python3.11/dist-packages/sklearn/utils/optimize.py:319: ConvergenceWarning: newton-cg failed to converge at loss = 0.3923257639127942. Increase the number of iterations.
  warnings.warn(


In [73]:
# prompt: print what parameter they used

print("Model Logistic best parameters:", model_logistic.best_params_)


Model Logistic best parameters: {'C': 0.1, 'max_iter': 500, 'penalty': 'l2', 'solver': 'newton-cg', 'tol': 1e-05}


## **d. Evaluasi Model Klasifikasi setelah Tuning (Optional)**

Berikut adalah rekomendasi tahapannya.
1. Gunakan model dengan hyperparameter terbaik.
2. Hitung ulang metrik evaluasi untuk melihat apakah ada peningkatan performa.

In [74]:
model_random_forest.predict(X_test)

model_random_forest.score(X_test, y_test)

model_logistic_param = LogisticRegression( C= 0.1,
                                           max_iter= 500,
                                           penalty= 'l2',
                                           solver= 'newton-cg',
                                           tol= 1e-05,

)

model_logistic_param.fit(X_train, y_train)

model_logistic_param.predict(X_test)


print("Model Random Forest")
print(classification_report(y_test, model_random_forest.predict(X_test)))
print(confusion_matrix(y_test, model_random_forest.predict(X_test)))
print("Accuracy Score : ", accuracy_score(y_test, model_random_forest.predict(X_test)))
print("F1 Score : ", f1_score(y_test, model_random_forest.predict(X_test), average='weighted'))

print("Model Logistic")
print(classification_report(y_test, model_logistic_param.predict(X_test)))
print(confusion_matrix(y_test, model_logistic_param.predict(X_test)))
print("Accuracy Score : ", accuracy_score(y_test, model_logistic_param.predict(X_test)))
print("F1 Score : ", f1_score(y_test, model_logistic_param.predict(X_test), average='weighted'))


Model Random Forest
              precision    recall  f1-score   support

           0       0.86      0.96      0.91        80
           1       0.99      1.00      1.00       147
           2       1.00      0.83      0.91         6
           3       1.00      0.99      0.99        92
           4       1.00      0.24      0.38        21
           5       1.00      0.93      0.96        14
           6       0.83      1.00      0.91         5
           7       0.97      0.97      0.97       116
           8       0.73      0.91      0.81        33

    accuracy                           0.94       514
   macro avg       0.93      0.87      0.87       514
weighted avg       0.95      0.94      0.94       514

[[ 77   0   0   0   0   0   0   0   3]
 [  0 147   0   0   0   0   0   0   0]
 [  0   0   5   0   0   0   1   0   0]
 [  0   0   0  91   0   0   0   1   0]
 [  9   0   0   0   5   0   0   1   6]
 [  0   1   0   0   0  13   0   0   0]
 [  0   0   0   0   0   0   5   0   0]
 [

/usr/local/lib/python3.11/dist-packages/sklearn/utils/optimize.py:319: ConvergenceWarning: newton-cg failed to converge at loss = 0.3923257639127942. Increase the number of iterations.
  warnings.warn(


## **e. Analisis Hasil Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Bandingkan hasil evaluasi sebelum dan setelah tuning (jika dilakukan).
2. Identifikasi kelemahan model, seperti:
  - Precision atau Recall rendah untuk kelas tertentu.
  - Apakah model mengalami overfitting atau underfitting?
3. Berikan rekomendasi tindakan lanjutan, seperti mengumpulkan data tambahan atau mencoba algoritma lain jika hasil belum memuaskan.

In [75]:
print("Model Random Forest")
print(classification_report(y_test, model_random_forest.predict(X_test)))
print(confusion_matrix(y_test, model_random_forest.predict(X_test)))
print("Accuracy Score : ", accuracy_score(y_test, model_random_forest.predict(X_test)))
print("F1 Score : ", f1_score(y_test, model_random_forest.predict(X_test), average='weighted'))

print("Model Logistic")
print(classification_report(y_test, model_logistic_param.predict(X_test)))
print(confusion_matrix(y_test, model_logistic_param.predict(X_test)))
print("Accuracy Score : ", accuracy_score(y_test, model_logistic_param.predict(X_test)))
print("F1 Score : ", f1_score(y_test, model_logistic_param.predict(X_test), average='weighted'))

Model Random Forest
              precision    recall  f1-score   support

           0       0.86      0.96      0.91        80
           1       0.99      1.00      1.00       147
           2       1.00      0.83      0.91         6
           3       1.00      0.99      0.99        92
           4       1.00      0.24      0.38        21
           5       1.00      0.93      0.96        14
           6       0.83      1.00      0.91         5
           7       0.97      0.97      0.97       116
           8       0.73      0.91      0.81        33

    accuracy                           0.94       514
   macro avg       0.93      0.87      0.87       514
weighted avg       0.95      0.94      0.94       514

[[ 77   0   0   0   0   0   0   0   3]
 [  0 147   0   0   0   0   0   0   0]
 [  0   0   5   0   0   0   1   0   0]
 [  0   0   0  91   0   0   0   1   0]
 [  9   0   0   0   5   0   0   1   6]
 [  0   1   0   0   0  13   0   0   0]
 [  0   0   0   0   0   0   5   0   0]
 [

# Kesimpulan
##Perbandingan Kinerja Model

- Random Forest memiliki akurasi lebih tinggi (94.36%) dibanding Logistic Regression (93.58%).

- F1-score juga lebih tinggi pada Random Forest (0.9363) dibanding Logistic Regression (0.9331).

- Model Random Forest unggul dalam menangani kelas yang memiliki distribusi data tidak merata.

## Analisis Performa Model

- Random Forest memiliki precision dan recall yang lebih stabil di berbagai kelas, meskipun beberapa kelas seperti kelas 4 memiliki recall yang sangat rendah (24%), yang menunjukkan model sulit mengenali kelas ini.

- Logistic Regression memiliki kelemahan pada kelas 5 dan 6, dengan recall yang lebih rendah (masing-masing 43% dan 80%). Ini menunjukkan model kesulitan mengenali beberapa kelas dengan data lebih sedikit.

## Potensi Overfitting atau Underfitting

-Random Forest cenderung lebih fleksibel, tetapi perlu dicek apakah ada overfitting, terutama pada kelas dengan jumlah sampel kecil.

- Logistic Regression mengalami masalah konvergensi (Newton-CG gagal mencapai titik optimal), yang bisa menandakan underfitting atau data yang tidak cukup kompleks untuk model ini.

## Rekomendasi Tindakan Lanjutan

### Jika hasil belum memuaskan, bisa dilakukan:

- Tuning hyperparameter (terutama pada Random Forest) untuk mengurangi kemungkinan overfitting.

- Menyeimbangkan dataset untuk meningkatkan performa model pada kelas yang kurang terprediksi dengan baik.

- Mencoba pendekatan lain, seperti Gradient Boosting atau Neural Networks, jika dataset lebih kompleks.

- Mengumpulkan lebih banyak data, terutama untuk kelas yang memiliki support rendah (contoh: kelas 2, 5, dan 6).